File: tenspiler/3d_artificial/python/elemwise_add.py
from typing import List


def elemwise_add(
    tensor_a: List[List[List[int]]], tensor_b: List[List[List[int]]]
) -> List[List[List[int]]]:
    result: List[List[List[int]]] = []

    # Each element of tensor_a is a matrix
    for i in range(len(tensor_a)):
        curr_matrix: List[List[int]] = []
        # Each element of tensor_a[0] is a list
        for j in range(len(tensor_a[0])):
            curr_lst: List[int] = []
            for k in range(len(tensor_a[0][0])):
                curr_lst.append(tensor_a[i][j][k] + tensor_b[i][j][k])
            curr_matrix.append(curr_lst)
        result.append(curr_matrix)

    return result

File: tenspiler/3d_artificial/cpp/elemwise_add.cc
#include <vector>
using namespace std;

vector<vector<vector<int>>> elemwise_add(
    vector<vector<vector<int>>> tensor3d_x,
    vector<vector<vector<int>>> tensor3d_y
) {
    vector<vector<vector<int>>> out;
    int m = tensor3d_x.size();
    int n = tensor3d_x[0].size();
    int o = tensor3d_x[0][0].size();

    for (int i = 0; i < m; i++) {
        vector<vector<int>> matrix;
        for (int j = 0; j < n; j++) {
            vector<int> lst;
            for (int k = 0; k < o; k++) {
                lst.push_back(tensor3d_x[i][j][k] + tensor3d_y[i][j][k]);
            }
            matrix.push_back(lst);
        }
        out.push_back(matrix);
    }

    return out;
}
